scripts/mempool.js

```diff
@@ -0,0 +1,248 @@
+
+/** An Unspent Transaction Output, used as Inputs of future transactions */
+class UTXO {
+    /**
+     * @param {Object} UTXO
+     * @param {String} UTXO.id - Transaction ID
+     * @param {String} UTXO.path - If applicable, the HD Path of the owning address
+     * @param {Number} UTXO.sats - Satoshi value in this UTXO
+     * @param {String} UTXO.script - HEX encoded spending script
+     * @param {Number} UTXO.vout - Output position of this transaction
+     * @param {Number} UTXO.height - Block height of the UTXO
+     * @param {Number} UTXO.status - UTXO status enum state
+     */
+    constructor({id, path, sats, script, vout, height, status} = {}) {
+        /** Transaction ID
+         * @type {String} */
+        this.id = id;
+
+        /** HD Path of the owning address
+         *  @type {String} */
+        this.path = path;
+
+        /** Satoshi value in this UTXO
+         *  @type {Number} */
+        this.sats = sats;
+
+        /** HEX encoded spending script
+         *  @type {String} */
+        this.script = script;
+
+        /** Output position of this transaction
+         *  @type {Number} */
+        this.vout = vout;
+
+        /** Block height of the UTXO
+         *  @type {Number} */
+        this.height = height;
+
+        /** UTXO status enum state
+         *  @type {Number} */
+        this.status = status;
+    }
+};
+
+/** A Mempool instance, stores and handles UTXO data for the wallet */
+class Mempool {
+    constructor() {
+        /** 
+         * An array of all known UTXOs
+         * @type {Array<UTXO>}
+         */
+        this.UTXOs = [];
+    }
+
+    /** The CONFIRMED state (UTXO is spendable) */
+    static CONFIRMED = 0;
+
+    /** The REMOVED state (UTXO was spent and will be removed soon) */
+    static REMOVED = 1;
+
+    /** The PENDING state (standard UTXO is in mempool, pending confirmation) */
+    static PENDING = 2;
+
+    /** The DELEGATED PENDING state (cold UTXO is in mempool, pending confirmation) */
+    static PENDING_COLD = 3;
+
+    /** The CONFIRMED DELEGATED state (cold UTXO is spendable) */
+    static DELEGATE = 4;
+
+    /** The REWARD state (UTXO is a reward from a stake or masternode) */
+    static REWARD = 5;
+
+    /** The block count, updated automatically by the network sync module
+     * @type {Number}
+     */
+    static blockCount;
+
+    /**
+     * Remove a UTXO after a set amount of time
+     * @param {Number} nBlocks - Estimated blocks to wait
+     * @param {UTXO} cUTXO - UTXO to remove
+     */
+     async removeWithDelay(nBlocks, cUTXO) {
+        await sleep(nBlocks * 60 * 1000);
+        this.removeUTXO(cUTXO);
+    }
+
+    /**
+     * Check if an exact UTXO match can be found in our wallet
+     * @param {UTXO} cNewUTXO - UTXO to search for
+     * @returns {Boolean} `true` or `false`
+     */
+    isAlreadyStored(cNewUTXO) {
+        for (const cUTXO of this.UTXOs) {
+            if (JSON.stringify(cUTXO) === JSON.stringify(cNewUTXO)) {
```

Comment:
Can't we just check if the `txid` and `vout` are the same?

Reply:
not really there could be UTXOs with same id same vout but different status, and sometimes we may want to differentiate them